sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala

```diff
@@ -0,0 +1,497 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.shims
+
+import ai.rapids.cudf.{BinaryOperable, ColumnVector, DType, RoundMode, Scalar}
+import com.nvidia.spark.rapids.{Arm, GpuBinaryExpression, GpuColumnVector, GpuScalar}
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, ImplicitCastInputTypes, NullIntolerant}
+import org.apache.spark.sql.types._
+
+object IntervalUtils extends Arm {
+
+  /**
+   * compute interval.toLong * num
+   */
+  def mulIntervalAndNum(interval: BinaryOperable, num: BinaryOperable): ColumnVector = {
+    interval match {
+      case intervalVector: ColumnVector =>
+        withResource(intervalVector.castTo(DType.INT64)) { intervalLong =>
+          intervalLong.mul(num)
+        }
+      case intervalScalar: Scalar =>
+        withResource(Scalar.fromLong(getLong(intervalScalar))) { intervalScalar =>
+          intervalScalar.mul(num)
+        }
+    }
+  }
+
+  /**
+   * throws exception if any value in `longCv` exceeds the int limits
+   */
+  def toIntWithCheck(longCv: ColumnVector): ColumnVector = {
+    withResource(longCv.castTo(DType.INT32)) { intResult =>
+      withResource(longCv.notEqualTo(intResult)) { notEquals =>
+        if (notEquals.any().getBoolean) {
+          throw new ArithmeticException("overflow occurs")
+        } else {
+          intResult.incRefCount()
+        }
+      }
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param leftLong  long cv or scalar
+   * @param rightLong long cv or scalar, will not be scalar if leftLong is scalar
+   * @param result    is the result of x * y
+   */
+  def checkMultiplyOverflow(leftLong: BinaryOperable, rightLong: BinaryOperable,
+      result: ColumnVector): Unit = {
+    leftLong match {
+      case leftVector: ColumnVector =>
+        checkMultiplyOverflowImp(rightLong, leftVector, result)
+      case _ =>
+        checkMultiplyOverflowImp(leftLong, rightLong.asInstanceOf[ColumnVector], result)
```

Comment:
How do we know `rightLong` is a column vector here ?

Reply:
If `left` and `right` are both `Scalar` will invoke the following method.

For more detail, see `GpuBinaryExpression`
```
  override def doColumnar(numRows: Int, intervalScalar: GpuScalar,
      numScalar: GpuScalar): ColumnVector = {
    withResource(GpuColumnVector.from(intervalScalar, numRows, interval.dataType)) { expandedLhs =>
      doColumnar(expandedLhs, numScalar)
    }
  }
```